File: 6M4W/ehowlqk/boj_30804.js
const fs = require("fs");
const input = fs
  .readFileSync("./6M4W/ehowlqk/boj_30804.txt")
  .toString()
  .trim()
  .split("\n");

let start = 0;
let end = 1;

const fruits = input[1].split(" ").map(Number);

let tanghuru = new Map();
tanghuru.set(fruits[0], 1);

let answer = 1;
while (end < Number(input[0])) {
  if (tanghuru.size < 3) {
    tanghuru.set(fruits[end], (tanghuru.get(fruits[end]) || 0) + 1);
    end++;
    if (tanghuru.size < 3) {
      answer = Math.max(answer, end - start);
    }
  } else {
    if (tanghuru.get(fruits[start]) > 1) {
      tanghuru.set(fruits[start], tanghuru.get(fruits[start]) - 1);
    } else {
      tanghuru.delete(fruits[start]);
    }
    start++;
  }
}

console.log(answer);

File: 6M4W/ehowlqk/boj_30804.py
n = int(input())
tanghuru = list(map(int, input().split()))

start, end = 0, 1
answer = 1
fruits = {tanghuru[0]: 1}   # 과일 종류별 개수

while end < n:
    # 과일 두 개 이하로 만들었을 때:
    if len(fruits) < 3:
        fruits[tanghuru[end]] = fruits.get(tanghuru[end], 0) + 1    # 딕셔너리에 추가
        end += 1    # end 증가
        answer = max(answer, end - start) if len(fruits) < 3 else answer    # end 증가 후 과일 종류가 2개 이하일 때만 최대 길이 갱신

    else:
        if fruits[tanghuru[start]] == 1:    # start의 과일이 한 개 뿐일 때: 딕셔너리에서 제거
            fruits.pop(tanghuru[start])
        else:
            fruits[tanghuru[start]] -= 1    # 그 외: 개수 -1만큼 감소
        start += 1  # start 증가

print(answer)
